third_party/chimp/include/byte_reader.hpp

```diff
@@ -0,0 +1,169 @@
+#pragma once
+
+#include <stdint.h>
+#include <iostream>
+#include "duckdb/common/fast_mem.hpp"
+
+namespace duckdb_chimp {
+
+// This class reads arbitrary amounts of bits from a buffer
+// If 41 bits are requested (5 bytes + 1 bit), we will read 6 bytes and increment the byte index by 6
+// With the assumption that the remainder of the last byte read is zero-initialized
+class ByteReader {
+private:
+	static constexpr uint8_t final_shifts[8] = {
+		0,
+		7,
+		6,
+		5,
+		4,
+		3,
+		2,
+		1
+	};
+public:
+	ByteReader() : buffer(nullptr), index(0) {
+
+	}
+public:
+	void SetStream(uint8_t* buffer) {
+		this->buffer = buffer;
+		index = 0;
+	}
+
+	template <class T>
+	T ReadValue() {
+		throw std::runtime_error("fixme read value");
```

Comment:
I think I put this line here myself, but this should probably be an internal exception :)